cpp/src/io/parquet/writer_impl.cu

```diff
@@ -862,6 +861,10 @@ void writer::impl::write_chunk(table_view const &table, pq_chunked_state &state)
     // increment num rows
     state.md.num_rows += num_rows;
   }
+  // Create table_device_view so that corresponding column_device_view data
+  // can be written into col_desc members
+  auto parent_column_table_device_view = table_device_view::create(table);
+  auto leaf_column_table_device_view   = table_device_view::create(table);
```

Comment:
this looks unused

Reply:
It is being used at a later point in init_page_fragments [here](https://github.com/rapidsai/cudf/blob/90b5da019c33dfcfd1c1685b02d305c8d8e2cb70/cpp/src/io/parquet/page_enc.cu#L2008).

Reply:
This seems to store leaf column pointer values from parent table into leaf table. But I don't understand the purpose. Is it not enough to store pointers into the parent table in EncColumnDesc?

Reply:
You mean to say that just have `parent_column_table_device_view` and for each list type column, get the leaf column and store the pointer to it in `EncColumnDesc`?

Reply:
This function does not just store column pointer values but also stores column_device_views in device memory.

EncColumnDesc::leaf_column pointer needs to point to an object in memory. The function `InitColumnDeviceViews` traverses from a given column in `parent_column_table_device_view` down to the level of the leaf via the `child` function. However just storing the pointer to the object obtained via this function is not enough because it expires after the function call. I am storing that leaf `column_device_view` object into `leaf_column_table_device_view` like this :

```
        column_device_view *leaf_col_ptr = leaf_column_table_device_view.begin() + i;
        *leaf_col_ptr                    = col;
```
and then setting EncColumnDesc::leaf_column

`        col_desc[i].leaf_column          = leaf_col_ptr;`

So the purpose of the second table_view is to create device memory where the leaf column can be placed because the `child` function of a column returns an object and not a pointer to its child.

```
  __device__ column_device_view child(size_type child_index) const noexcept
  {
    return d_children[child_index];
  }
```
If there is a method to return the pointer instead then I think I can avoid creating `leaf_column_table_device_view`.

Reply:
> I am storing that leaf `column_device_view` object into `leaf_column_table_device_view`

I understood this. 

> the child function of a column returns an object and not a pointer to its child.

This, I didn't know. Pretty inconvenient but I suppose we cannot add a method to return a pointer to the children because the children layout is abstracted away.

However, what if we just stored the leaf `column_device_view` by value in `EncColumnDesc`? It's supposed to be cheap to copy and parent device table should be able to provide us the storage to keep it alive in terms of child pointers (string column would still have children despite being leaf)

Reply:
> However, what if we just stored the leaf column_device_view by value in EncColumnDesc?

We could do that (and keep parent column as pointer), but that would cause `EncColumnDesc` to not be default constructable anymore since `column_device_view` doesn't have a default constructor.
So lines like this won't be possible :
https://github.com/rapidsai/cudf/blob/fc40c52a898181d3b4e019134e3ea0b32841515e/cpp/src/io/parquet/writer_impl.cu#L878

I am not clear how we can have initialization of `hostdevice_vector` with a non default constructor.

We could create an equivalent `hostdevice_buffer` (is that a thing) which does not call a constructor.

Reply:
Understood. So certain unrelated limitations make this a necessary evil. Since we don't use all the storage allocated by `table_device_view::create(table)`, and don't need the storage initialised, can we use `device_uvector<column_device_view> leaf_column_vector`? That'd save an unnecessary memcpy and make the intent just slightly more clear.

Reply:
There was a whole discussion on Slack on whether `hostdevice_vector` should initialize the buffers. I can't find if we managed to agree on the solution, but one option I like is to make the (zero)initialization optional, since we often don't use the initial values. 
Also, `hostdevice_vector` is definitely overused. maybe it should not be used `EncColumnDesc` in the first place (I'll take a look later).

Reply:
@vuule, some of the problems we have with hd_vec could be solved by not using pinned memory for it since allocating it is so expensive. As of now we need to be cautious to only allocate one kind of hd_vec only once during the call (as opposed to once per column/chunk/page). Having pinned memory pool in RMM can solve this, and so can replacing the device storage of hd_vec with device_buffer/device_uvector.

Reply:
True, but I think it's overused beyond that. For example, I don't think it needs to be used when we are transferring data between host and device (or vice versa) only once. If that's the case here, we might as well use `device_uvector` and avoid the default constructor issue.